### PR TITLE
Adding ResultStorageEngine Support in AsyncQueryThread

### PR DESCRIPTION
**Pull request by moizarafat**
Co-authored-by: chandrakaran

Resolves  #1272

## Description
Adding ResultStorageEngine usage in AsyncQueryThread for storing the results.

## Motivation and Context
Refer #1272

## How Has This Been Tested?
Existing Test cases pass and new test cases.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 1]
```diff
@@ -45,53 +56,195 @@
...
         queryResultObj = new AsyncQueryResult();
         queryResultObj.setHttpStatus(response.getResponseCode());
-        queryResultObj.setResponseBody(response.getBody());
-        queryResultObj.setContentLength(response.getBody().length());
+        queryResultObj.setResponseBody(responseBody);
+        queryResultObj.setContentLength(responseBody.length());
         queryResultObj.setCompletedOn(new Date());
 
+        if (isError) {
+            queryResultObj.setResponseBody(responseBody);
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:123]: We already did this on line 118.

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 2]
```diff
@@ -45,53 +56,195 @@
...
+            return queryResultObj;
+        }
+
+        recCount = calculateRecordCount(response, queryType);
+        queryResultObj.setRecordCount(recCount);
+
+        if (resultFormatType == ResultFormatType.CSV) {
+            observableResult = convertJsonToCSV(responseBody);
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:131]: This conversion is not save to do in memory.  We do this before checking if the type is download or not.

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 3]
```diff
@@ -45,53 +56,195 @@
...
+        }
+
+        if (isDownload) {
+            queryObj.setResult(queryResultObj);
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:135]: Why do we only set the result for downloads?

File: elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java [hunk 4]
```diff
@@ -45,53 +56,195 @@
...
+        List<Object[]> json2csv;
+
+        try {
+            json2csv = flatMe.json2Sheet().headerSeparator("_").getJsonAsSheet();
+        } catch (Exception e) {
+            log.debug("Exception while converting to CSV: {}", e.getMessage());
+            throw new IllegalStateException(e);
+        }
+
+        return Observable.using(
```
Comment (aklish) [inline elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java:204]: This doesn't look right to me.   We shouldn't be handling the entire response body here and convert it to CSV.  We should only convert one row at a time.

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-integration-tests/src/test/java/com/yahoo/elide/async/integration/tests/AsyncIT.java [hunk 5]
```diff
@@ -389,6 +390,91 @@ public void graphQLHappyPath1() throws InterruptedException {
...
         }
     }
+
+    /**
+     * Test for a GraphQL query as a Async Request with asyncAfterSeconds value set to 0.
+     * Happy Path Test Scenario 1 when resultType is DOWNLOAD
+     * @throws InterruptedException
+     */
+    @Test
+    @Tag("skipInMemory") //Without an ORM, there is nothing to hydrate the excluded entity
```
Comment (moizarafat) [inline elide-integration-tests/src/test/java/com/yahoo/elide/async/integration/tests/AsyncIT.java:400]: remove this, as we dont have DB dependency for result storage engine anymore.

File: elide-graphql/src/main/java/com/yahoo/elide/graphql/QueryRunner.java [hunk 6]
```diff
@@ -203,7 +280,7 @@ private ElideResponse executeGraphQLRequest(String baseUrlEndPoint, ObjectMapper
             tx.preCommit();
             requestScope.runQueuedPreSecurityTriggers();
             requestScope.getPermissionExecutor().executeCommitChecks();
-            if (query.trim().startsWith(MUTATION)) {
+            if (isMutation(query)) {
```
Comment (moizarafat) [inline elide-graphql/src/main/java/com/yahoo/elide/graphql/QueryRunner.java:283]: graphQLQuery.isMutation can be used.

**Pull request closed without merging**